### PR TITLE
BNS Model

### LINKED ISSUE #125: Integrate BNS model training code

### LINKED ISSUE #139: Add checkpointing to training `cli`

### LINKED ISSUE #140: Generalize `ModelCheckpoint` jit tracing for frequency and time domain use cases
Currently the `ModelCheckpoint` will fail at tracing the model when using spectrogram inputs since it [assumes](https://github.com/ML4GW/aframev2/blob/0ec45a39afab6a1adac4b33da00b98cf31c51d75/projects/train/train/callbacks.py#L25) the input is 1D. 

We can generalize this by inferring whether we are in the 2D or 1D regime by inspecting the `pl_module.model`:
```python
def on_train_end(self, trainer, pl_module):
    ...
    if isinstance(pl_module.model, SupervisedTimeDomainArchitecture):
        sample_input = torch.randn(1, datamodule.num_ifos, kernel_size)
    else:
        sample_input = torch.randn(1, datamodule.num_ifos, dim1, dim2)
     ...
```

Or, we can infer the input size directly from the dataloader:
```python
def on_train_end(self, trainer, pl_module):
    ...
    batch = next(trainer.datamodule.train_dataloader())
    batch = trainer.datamodule.augment(batch)
    sample_input = batch[0]
    ...
```

### LINKED ISSUE #142: Circular imports in `projects/train/train/data`
The `__init__.py` in `projects/train/train/data` has `from .supervised import *`. In `supervised`, the `__init__.py` has 
```
from .supervised import SupervisedAframeDataset
from .time_domain import TimeDomainSupervisedAframeDataset
```
And in `time_domain.py`, we have `from train.data.supervised import SupervisedAframeDataset`. This all works because `SupervisedAframeDataset` is imported before `TimeDomainSupervisedAframeDataset`. However, `pre-commit` puts `from.frequency_domain import FrequencyDomainSupervisedAframeDataset` _before_ the `SupervisedAframeDataset` import. This leads to a circular import.

I think the options are to either add another directory level, or tell `pre-commit` to skip that line. Am I missing a nicer option?

### PR DESCRIPTION
**Pull request by wbenoit26**
Adds the infrastructure necessary to do training using Q-transforms, with the `SingleQTransform` module from `ml4gw.
- Adds BNS prior based on R&P injections
- Adds args for the Q-transform into the train config
- Saving the best model now uses the `dataloader` to infer model input size
- Adds ability to specify a checkpoint from which to restart training
- For `export`, model input size is inferred from the training batch saved during training
- Adds an `augmentor` to the `BatchWhitener`, which acts on the whitened data before the data is passed to the NN
- Implements workaround for SV plotting for a single mass combination, and makes `mass_combo` an argument of the function

Closes #125 
Closes #139 
Closes #140 
Closes #142

### CODE REVIEW

**Comment by wbenoit26** [comment]
@EthanMarx Curious to get your thoughts here. I don't love the way that `q` gets handled, and more generally, how any future `augmentor` args would have to be passed. I'm also thinking that it would be a good idea to make a place to put different configs for different scenarios, so that they can be easily swapped out without needing to change each parameter manually. We could have a  set of base BBH configs and a set of base BNS configs that could be further modified for whatever experiment we're running.

**Comment by EthanMarx** [comment]
Looking at the configuration changes, I think that multiple config idea is necessary now. Could you add a `configs` directory at the root of the `train` project? We can then have `bns.yaml` `bbh.yaml` and `tune.yaml` live there.

File: projects/train/train/callbacks.py
Comment (EthanMarx) [inline projects/train/train/callbacks.py:21]: Glad this worked. Just wan't to confirm that this doesn't break the BBH case. Could you double check this? (we should probably find a way to incorporate this into tests!)

File: projects/train/train/data/supervised/frequency_domain.py
Comment (EthanMarx) [inline projects/train/train/data/supervised/frequency_domain.py:23]: Can you explain why this is need here?

File: projects/utils/utils/preprocessing.py
Comment (EthanMarx) [inline projects/utils/utils/preprocessing.py:107]: Missing `Optional` type hint

File: projects/train/train/data/supervised/frequency_domain.py
Comment (wbenoit26) [reply projects/train/train/data/supervised/frequency_domain.py:23]: Because `X_fg` is [4D tensor](https://github.com/ML4GW/aframev2/blob/380921c48bf593a41e81298f4b7abd3b84f60c0a/projects/train/train/data/base.py#L467), and the `qtransform` requires 3 or fewer dimensions. This is what I was trying to solve with that `ml4gw` PR, but the interpolation issues don't seem to permit any nice solutions.

**Comment by EthanMarx** [comment]
TODOs from our call:

1. Move `qtransform` module into the `forward` method of a `pl.LightningModule` so that it can be traced with the model at export time. Maybe generalize this to some `augmentor = Optional[torch.nn.Module]` and then it can be configured in the `yaml` config
2. Remove `q` parameter from `luigi` pipeline. This will now live in the training `config.yaml`
3. Export code no longer needs to infer input size from batch file, or include the `qtransform` in the `BatchWhitener`. I think there shouldn't be any changes to the export code.
4. Test out training pipeline for BBH and BNS to ensure that the model tracing during checkpointing is completely general.

File: projects/train/train/callbacks.py
Comment (wbenoit26) [reply projects/train/train/callbacks.py:24]: Confirmed that both the BNS and BBH case work

**Comment by wbenoit26** [comment]
@EthanMarx I was able to incorporate the `augmentor` into the model. It's a little less neat than I was hoping, but it works for both BNS and BBH cases. One thing I'm not loving is that during training, the log message is
```
  | Name      | Type           | Params
---------------------------------------------
0 | augmentor | Identity       | 0
1 | model     | Sequential     | 7.2 M
2 | metric    | TimeSlideAUROC | 0
---------------------------------------------
```
Rather than something more specific for the model. Not sure if there's a way to show the architecture there, rather than the model.

I'm realizing that this change kind of removes the need to have separate time and frequency domain data objects, so we may want to deprecate them in the future unless we find another need.

The only remaining issue is with exporting the Q-transform because of the FFTs. It looks like PyTorch has a beta `dynamo_export` that can do the conversion, but I need to play around with it to see if all the version changes will work.

**Comment by EthanMarx** [comment]
Ah yeah for clarity sake, it be worth removing the `torch.nn.Sequential` and just calling the augmentor by hand. 

Also, good point on the `FrequencyDomainDataset`, if it's not being used anywhere, I would say remove it, or open an issue to remove it down the line. 

Is it not currently possible to export the ffts? If so, we should probably reconsider this setup

**Comment by wbenoit26** [comment]
If I call the augmentor by hand, then the trace won't include it, and it won't get saved with the rest of the model.

I can't export as-is, but ONNX supports FFTs, it's just a matter of whether the converter between PyTorch and ONNX has those included. It looks like it's possible, but might require changing so much that it's not worth it. I'm going to see what it takes to get things working.

**Comment by EthanMarx** [comment]
I think as long as you call the `augmentor` in the forward method it should get traced

**Comment by wbenoit26** [comment]
I tried that, and I ran into issues loading the model from checkpoint to do the trace. But now that I'm thinking about it again, I think I know what I was doing wrong. I'll re-attempt.

**Comment by EthanMarx** [comment]
Hoping this is as simple as replacing `torch.onnx.export` with `torch.onnx.dynamo_export` in `hermes`. It looks like `dynamo_export` is still in beta version, but it might be stable enough to support our few use cases. 

We can add some flag `dynamo` boolean flag to the [`TorchOnnx`](https://github.com/ML4GW/hermes/blob/main/hermes/quiver/exporters/torch_onnx.py#L31) exporter that will use whichever one is desired.

**Comment by wbenoit26** [comment]
It also requires [onnxscript](https://onnxscript.ai), which, judging by the [PyPI release history](https://pypi.org/project/onnxscript/#history), is also very much in development. Again, might be stable enough for us.

It turns out that `dynamo` and `jit` are separate frameworks, so more changes would be necessary. It seems like `dynamo` will be what's used more going forward, so I think it'll be worth it to figure out. I'll make a separate PR for just those changes and then rebase this one.

**Comment by wbenoit26** [comment]
@EthanMarx As we discussed, I've reverted back to having the Q-transform explicitly in `export`. Including as part of the model will be the right thing to do eventually, but that requires a possibly significant update to `hermes`, and that shouldn't hold up this PR.

I've removed the `q` parameter from the pipeline config and the tasks, so it just lives in the `train` and `export` `config.yaml` files. I think it's cleaner this way, even though there's nothing to enforce that the value in each project be the same.

I've tested both the `bbh.yaml` and the `bns.yaml` config files (with `max_epochs = 2`), and both are able to get through training and export. Let me know if you have further thoughts on this PR.

File: aframe/tasks/plots/sv.py
Comment (EthanMarx) [inline aframe/tasks/plots/sv.py:32]: Maybe a description like "Python path to prior that was used for generating testing waveform injections" will be a bit more clear

**Comment by wbenoit26** [comment]
Ah okay, I just assumed that was set up already. I'm not sure how it was working if `None` was being passed for `q`, but maybe some default value was being used (or maybe I missed something in testing). Was there a reason that the `config.yaml` wasn't being used yet, or was it just still on the to-do list?

**Comment by EthanMarx** [comment]
To use the `config.yaml` we would need to start calling the export `cli` directly from the luigi task. Currently, I just import and call the executable itself. 

The main reason I did this was to avoid having to manage multiple configs that share variables, and to avoid too much confusion over which variables are being passed where. This works great when the `train` and `export` projects are entirely decoupled, but it seems that's no longer going to be the case. 

I think there must be a way for the export config to import/share variables from the train config, in which case I think this problem could be solved. Let me look into this.

**Comment by wbenoit26** [comment]
That would be great. I was wondering if it would ever make sense to combine the train and export projects, but being able to import the config file would be a much neater solution.

**Comment by EthanMarx** [comment]
Yeah thats definitely an idea worth considering at this point

File: projects/train/configs/bns.yaml
Comment (wbenoit26) [inline projects/train/configs/bns.yaml:62]: @EthanMarx Is there a way to have this value use the value for `kernel_length` given in the sandbox config? I.e., can multiple arguments in this config be taken from a single argument in the sandbox config?

Comment (EthanMarx) [reply projects/train/configs/bns.yaml:62]: The way to do this would be via [argument linking](https://lightning.ai/docs/pytorch/stable/cli/lightning_cli_expert.html#argument-linking) in the LightningCLI parser. 

Since this won't be needed in the BBH case, might make sense to subclass the CLI for this.

**Comment by wbenoit26** [comment]
@EthanMarx I went back to having `q` specified in the sandbox config, and I realized that there's an `OptionalFloatParameter` that makes things neater, though still not perfect. I also found that things worked a little more smoothly if I had the Q-transform be instantiated in the `setup` function of the dataset, so I switched to doing things that way.

Let me know what you think.

**Comment by EthanMarx** [comment]
Yeah that definitely makes sense for now. Looks good. Is this ready for a final once over?

**Comment by wbenoit26** [comment]
Yeah, I'd say so.

File: aframe/pipelines/sandbox/sandbox.cfg
Comment (EthanMarx) [inline aframe/pipelines/sandbox/sandbox.cfg:1]: Should we just get rid of this now that we have a configs directory?

File: projects/export/export/snapshotter.py
Comment (EthanMarx) [inline projects/export/export/snapshotter.py:41]: Would a cleaner/more robust safeguard here be something like 

```python
If q is not None:
    if len(kernel_size) != 2:
        raise ValueError("Something is wrong")
    augmentor = ...
```

File: projects/train/train/data/base.py
Comment (EthanMarx) [inline projects/train/train/data/base.py:72]: Main reason I left the type hint so general was to support our old `SnrSampler` which I don't think jsonargparse would treat as compatible here

**Review by EthanMarx — APPROVED** [review]
@wbenoit26 Left a few other relatively minor comments. Don't want to hold this up for too much longer so will approve and let you decide what's worthy. Feel free to just open up issues for unresolved items

File: aframe/pipelines/sandbox/sandbox.cfg
Comment (wbenoit26) [reply aframe/pipelines/sandbox/sandbox.cfg:1]: No, I don't think so. I like the idea of having a "working config", which you modify during whatever testing/analysis you want to do, and a set of "basis configs", which get changed much less frequently and that you can copy over to `sandbox.cfg` to start running things. That way, you don't have to worry about a PR accidentally changing something significant in the config and needing to double-check things after pulling the latest version.

File: projects/export/export/snapshotter.py
Comment (wbenoit26) [reply projects/export/export/snapshotter.py:41]: That's a good idea, I'll update

File: projects/train/train/data/base.py
Comment (wbenoit26) [reply projects/train/train/data/base.py:72]: True, I can include the other types as well

**Comment by wbenoit26** [comment]
@EthanMarx Looks like pushing additional commits dismissed your review, can you re-approve?

**Review by EthanMarx — APPROVED** [review]

**Pull request merged**
